server/chat-plugins/lottery.js

```diff
@@ -0,0 +1,248 @@
+'use strict';
+
+const LOTTERY_FILE = 'config/chat-plugins/lottery.json';
+
+/** @type {typeof import('../../lib/fs').FS} */
+const FS = require(/** @type {any} */('../../.lib-dist/fs')).FS;
+
+/** @type {{[roomid: string]: {maxWinners: number, name: string, markup: string, participants: {[ip: string]: string}}}} */
+let lotteries = {};
+try {
+	lotteries = require(`../../${LOTTERY_FILE}`);
+} catch (e) {
+	if (e.code !== 'MODULE_NOT_FOUND' && e.code !== 'ENOENT') throw e;
+}
+if (!lotteries || typeof lotteries !== 'object') lotteries = {};
+
+class Lottery {
+	/**
+     * @param {string} roomid
+     */
+	constructor(roomid) {
+		this.roomid = roomid;
+	}
+	/**
+	 * @param {number} maxWinners
+	 * @param {string} name
+	 * @param {string} markup
+	 */
+	create(maxWinners, name, markup) {
+		if (!lotteries[this.roomid]) {
+			lotteries[this.roomid] = {maxWinners, name, markup, participants: {}};
+			this.write();
+		}
+	}
+	write() {
+		FS(LOTTERY_FILE).writeUpdate(() => JSON.stringify(lotteries));
+	}
+	destroy() {
+		delete lotteries[this.roomid];
+		this.write();
+	}
+	/**
+     * @param {User} user
+     */
+	addUser(user) {
+		if (!lotteries[this.roomid].participants[user.latestIp]) {
+			lotteries[this.roomid].participants[user.latestIp] = user.name;
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	/**
+     * @param {User} user
+     */
+	removeUser(user) {
+		if (lotteries[this.roomid].participants[user.latestIp]) {
+			delete lotteries[this.roomid].participants[user.latestIp];
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	getWinner() {
+		const winners = [];
+		const lotteryData = lotteries[this.roomid];
+		let participantsKeys = Object.keys(lotteryData.participants);
+		for (let i = 0; i < lotteryData.maxWinners; i++) {
+			const randomKey = participantsKeys[participantsKeys.length * Math.random() << 0];
+			const winner = lotteryData.participants[randomKey];
+			winners.push(winner);
+			participantsKeys = participantsKeys.filter(username => username !== winner);
+		}
+		return winners;
+	}
+}
+
+/** @type {ChatCommands} */
+exports.commands = {
+	lottery: {
+		'!': true,
+		''(target, room) {
+			if (!room) {
+				return this.errorReply("This can only be used in rooms.");
+			}
+			const lotteryData = lotteries[room.id];
+			if (!lotteryData) {
+				return this.errorReply("This room doesn't have a lottery running.");
+			}
+			return this.parse(`/join view-lottery-${room.id}`);
+		},
+		'!create': true,
+		create(target, room, user) {
+			if (!this.can('declare', null, room)) return;
+			if (!room) {
+				return this.errorReply("This can only be used in rooms.");
+			}
+			if (lotteries[room.id]) {
+				return this.errorReply(`There is already a lottery running.`);
+			}
+			const lottery = new Lottery(room.id);
+			let [maxWinners, name, ...markup] = target.split(',').map(val => val.trim());
```

Comment:
It doesn't appear that ``markup`` ever gets ran past ``this.canHTML(foo)``, unless I'm mistaken